readme.md
# WEBVIEW GOLANG

It uses Cocoa/WebKit on macOS, gtk-webkit2 on Linux and Edge on Windows 10.

## USAGE

**GOLANG**

```
webview [options]

  --dir string
        path to serve (default ".")
  --url string
        instead of serving files, load this url
  --title string
        title of the webview window (default "webview")
  --width int
        width of the webview window (default 800)
  --height int
        height of the webview window (default 600)
```

**Nodejs**

```
npm install --global .
```

```
webview [options]

  --dir string
        path to serve (default ".")
  --url string
        instead of serving files, load this url
  --title string
        title of the webview window (default "webview")
  --width int
        width of the webview window (default 800)
  --height int
        height of the webview window (default 600)
```

**Deno**

```
deno install -f -A webviewd.ts
```

```
webviewd [options]

  --dir string
        path to serve (default ".")
  --url string
        instead of serving files, load this url
  --title string
        title of the webview window (default "webview")
  --width int
        width of the webview window (default 800)
  --height int
        height of the webview window (default 600)
```

## Distributing webview apps

On Linux you get a standalone executable. It will depend on GTK3 and GtkWebkit2,
so if you distribute your app in DEB or RPM format include those dependencies.
An application icon can be specified by providing a .desktop file.

```linux
sudo apt-get install libwebkit2gtk-4.0-dev
```

## GENERATE DEBIAN PACKAGE

**ref:** http://sdz.tdct.org/sdz/creer-un-paquet-deb.html

**red:** https://github.com/practice-golang/hello-cmake

**ref:**
https://stackoverflow.com/questions/61507209/creating-a-go-binary-as-debian-binary-package-for-a-custom-repository

**ref:** https://gowebexamples.com/templates/

## INSTALL AND REMOVE DEB

```
sudo dpkg -i nom_du_paquet.deb
sudo apt-get remove nom_du_paquet
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.10](https://github.com/stephendltg/go-webview/compare/v0.0.9...v0.0.10) (2021-08-16)

### [0.0.9](https://github.com/stephendltg/go-webview/compare/v0.0.8...v0.0.9) (2021-08-16)


### Features

* **darwin:** add darwin version for go/nodejs/deno ([8b2ee26](https://github.com/stephendltg/go-webview/commit/8b2ee268b08daa06938ace1783b62e743c846553))

### [0.0.8](https://github.com/stephendltg/go-webview/compare/v0.0.7...v0.0.8) (2021-08-04)

### [0.0.7](https://github.com/stephendltg/go-webview/compare/v0.0.6...v0.0.7) (2021-08-04)


### Features

* **bundle:** add win32 support ([d5b54f4](https://github.com/stephendltg/go-webview/commit/d5b54f442e07630b7d2bceb7e4f8ad8e074ce8ae))

### [0.0.6](https://github.com/stephendltg/go-webview/compare/v0.0.5...v0.0.6) (2021-08-04)

### [0.0.5](https://github.com/stephendltg/go-webview/compare/v0.0.4...v0.0.5) (2021-08-04)


### Features

* **deno:** add deno method ([59bfd99](https://github.com/stephendltg/go-webview/commit/59bfd993d5109499d5e42f253df579a814ac03ef))


### Bug Fixes

* **bundle:** change dir bin ([7b1a594](https://github.com/stephendltg/go-webview/commit/7b1a5945cc0e0ece7914303ec2ab4c2ffb93fdeb))
* **deno:** add dirname in path ([b82c379](https://github.com/stephendltg/go-webview/commit/b82c379fb78bc3e6bb8631de0441af8a66c691a7))

### [0.0.4](https://github.com/stephendltg/skeleton-go-webview/compare/v0.0.3...v0.0.4) (2021-08-03)

### [0.0.3](https://github.com/stephendltg/skeleton-go-webview/compare/v0.0.2...v0.0.3) (2021-08-03)

### Features

- **main.go:** add url or path for webview
  ([d6fe404](https://github.com/stephendltg/skeleton-go-webview/commit/d6fe40427886d0f8586ae47f0c62089a06da2dea))

### 0.0.2 (2021-08-03)

### Bug Fixes

- **git:** add hook git
  ([303a314](https://github.com/stephendltg/skeleton-go-webview/commit/303a314f79e2ddf7d6c484fe6faecfd9ecbd124d))


webviewd.ts
#!/usr/bin/env deno run --allow-env --allow-run

import { join } from "https://deno.land/std@0.103.0/path/mod.ts";
import * as log from "https://deno.land/std@0.103.0/log/mod.ts";

const __dirname = new URL(".", import.meta.url).pathname;

var binaryPath;

if (Deno.build.os === "windows") {
  binaryPath = join(__dirname, "bin", "webview-win32-amd64.exe").slice(1);
} else if (Deno.build.os === "darwin") {
  binaryPath = join(__dirname, "bin", "webview-darwin-amd64");
} else if (Deno.build.os === "linux") {
  binaryPath = join(__dirname, "bin", "webview-linux-amd64");
} else {
  log.critical("⚠ Unsupported platform: " + Deno.build.os);
}

if (!binaryPath) {
  Deno.exit(0);
}

// create webview
const webview = Deno.run({
  cmd: [binaryPath, ...Deno.args],
  stdout: "piped",
  stderr: "piped",
});

// await its completion
const { code } = await webview.status();

const rawOutput = await webview.output();
const rawError = await webview.stderrOutput();

if (code === 0) {
  await Deno.stdout.write(rawOutput);
} else {
  const errorString = new TextDecoder().decode(rawError);
  console.log(errorString);
}

Deno.exit(code);


main.go
package main

import (
	"flag"
	"fmt"
	"net"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"github.com/webview/webview"
)

// Vars cli
var (
	title  = flag.String("title", "default", "Webview title")
	debug  = flag.Bool("debug", false, "Mode debug")
	height = flag.Int("height", 600, "Webview height")
	width  = flag.Int("width", 800, "Webview width")
	url    = flag.String("url", "", "Naviagte to this url")
	dir    = flag.String("dir", ".", "path to serve")
)

// Init
func init() {

	// Parse cli
	flag.Parse()

}

// Get port available
func getport() (int, net.Listener) {
	listener, err := net.Listen("tcp", ":0")
	if err != nil {
		panic(err)
	}

	port := listener.Addr().(*net.TCPAddr).Port

	return port, listener
}

// Handle http server
func handle(fs http.Handler) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Add("Referrer-Policy", "0")
		// Cache
		w.Header().Add("Surrogate-Control", "no-store")
		w.Header().Add("Cache-Control", "no-store, no-cache, must-revalidate, proxy-revalidate")
		w.Header().Add("Pragma", "no-cache")
		w.Header().Add("Expires", "0")
		// Security
		w.Header().Add("X-Content-Type-Options", "nosniff")
		w.Header().Add("X-DNS-Prefetch-Control", "off")
		w.Header().Add("X-Frame-Options", "DENY")
		w.Header().Add("Strict-Transport-Security", "5184000")
		w.Header().Add("X-Download-Options", "noopen")
		w.Header().Add("X-XSS-Protection", "1; mode=block")

		fs.ServeHTTP(w, r)
	}
}

// static server
func fileServer(listener net.Listener, path string) {
	http.Handle("/", handle(http.FileServer(http.Dir(path))))
	panic(http.Serve(listener, nil))
}

// Absolu path exec
func abspath() string {
	exe, err := os.Executable()
	if err != nil {
		panic(err)
	}
	exe, err = filepath.EvalSymlinks(exe)
	if err != nil {
		panic(err)
	}
	dir := filepath.Dir(exe)
	return dir
}

// Run webview
func runWebview(url string, title string, width int, height int, debug bool) {
	w := webview.New(debug)
	defer w.Destroy()
	w.SetTitle(title)
	w.SetSize(width, height, webview.HintNone)
	w.Navigate(url)
	// Func close webview for js
	w.Bind("close", func() {
		os.Exit(0)
	})
	w.Run()
}

func main() {

	port, listener := getport()

	var path string
	if strings.HasPrefix(*dir, "./") || strings.HasPrefix(*dir, "../") || strings.HasPrefix(*dir, ".\\") || strings.HasPrefix(*dir, "..\\") {
		path = filepath.Join(abspath(), *dir)
	} else {
		path = *dir
	}

	if *url == "" {
		go fileServer(listener, path)
		url := fmt.Sprintf("http://localhost:%d/", port)
		runWebview(url, *title, *width, *height, *debug)
	} else {
		runWebview(*url, *title, *width, *height, *debug)
	}
}
